crates/miden-lib/asm/miden/contracts/auth/multisig.masm

```diff
@@ -1,24 +1,244 @@
+use.std::sys
 
 # CONSTANTS
 # =================================================================================================
 
-# Event to place the falcon signature of a provided message and public key on the advice stack.
-const.FALCON_SIG_TO_STACK=131087
+# The event to request an authentication signature.
+const.AUTH_REQUEST=131087
+
+# The event emitted when a signature is not found for a required signer.
+const.UNAUTHORIZED_EVENT=131100
+
+# The slot in this component's storage layout where the signature threshold is stored.
+const.THRESHOLD_SLOT=0
+
+# The slot in this component's storage layout where the number of approvers is stored.
+const.NUMBER_OF_APPROVERS_SLOT=1
+
+# The slot in this component's storage layout where the public keys map is stored.
+const.PUBLIC_KEYS_MAP_SLOT=2
+
+# The slot in this component's storage layout where executed transactions are stored.
+const.EXECUTED_TXS_SLOT=3
+
+# Runtime Memory Addresses
+const.TEMP_PK=108
+const.MESSAGE=112
+const.OWNER_PUB_KEY=120
+
+# ERRORS
+const.NON_CONSTANT_MSG="NON CONSTANT SIGNATURES"
+const.INVALID_SIGNATURE="SIG DOES NOT MATCH DIGEST"
+
+#! Verify signatures for all required signers in a loop.
+#!
+#! This procedure iterates through the required number of signers, fetches their public keys
+#! from the account storage map, and verifies their signatures against the transaction message.
+#! If any required signature is missing, it emits an UNAUTHORIZED_EVENT.
+#!
+#! Inputs:  [num_required_signatures]
+#! Outputs: []
+#!
+#! Invocation: exec
+proc.verify_signatures_loop.1
+    sub.1
+    # => [num_required_signatures-1]
+
+    # Counter `i` starts at `num_required_signatures-1` and counts down to 0
+    # => [i]
+
+    # Loop through required signers and verify signatures.
+    # ---------------------------------------------------------------------------------------------
+
+    dup neq.0
+    while.true
+        # => [i]
+
+        push.111 debug.stack drop
+
+        # Fetch public key from storage map.
+        # -----------------------------------------------------------------------------------------
+
+        dup sub.1 push.0.0.0 push.PUBLIC_KEYS_MAP_SLOT
+        # => [PUBLIC_KEYS_MAP_SLOT, [0, 0, 0, i-1], i]
+
+        exec.account::get_map_item
+        # => [OWNER_PUB_KEY, i]
+
+        mem_storew.TEMP_PK
+        # => [OWNER_PUB_KEY, i]
+
+        # Check if signature exists for this signer.
+        # -----------------------------------------------------------------------------------------
+
+        padw mem_loadw.MESSAGE
+        # => [MESSAGE, OWNER_PUB_KEY, i]
+
+        hmerge
+        # => [SIG_KEY, i]
+
+        adv.has_mapkey
+        # => [SIG_KEY, i]
+
+        adv_push.1 not
+        # => [unauthorized, SIG_KEY, i]
+
+        # If signatures are non-existent the tx will fail here.
+        if.true
+            mem_loadw.MESSAGE
+            # => [MSG, SIG_KEY, i]
+
+            emit.UNAUTHORIZED_EVENT
+        end
+        # => [SIG_KEY, i]
```

Comment:
The current approach will end up failing in the "success" case where the i-th signature is not provided, but where we overall reach the threshold.